### DISCUSSION TITLE
Modulation-Matrix and LFO - not only for MOD-Wheel

### DISCUSSION THREAD

**Discussion by ErichHeinemann** [marcel-licence/esp32_basic_synth#37] [Ideas]
Hi Marcel,
last note for today. :)

In my old code, I tried to mimic a Modulation-Matrix, - or started to do so.
In general, one or two slow LFOs could modulate a lot of things. directly 1:1 or by 50% (0.5f) or by another Multiplier 0.0 - 1.0  .. etc.
The Velocity could be one part of that multiplier for the level of the modulation.
In big Matrix-Synths, a lot of things could be part of this multipler, - the LFOs too or any Controller.
The Destination of this LFO-Values and other Multipliers could be nearly everything - Pitch, ADSR, Filter, etc.
With that, every note could be slightly different to the note before and makes the sound more natural.

The modulation-wheel, or other CCs could be used to add more Value to the Modulation-Matrix.
I believe, Your Roland FA has a Modulation-Matrix too. :)

I tried Your new synth-version today in UNISON-Mode with  a slightly modified detune and more voice and it was really fat.
With some more "random" Modulation, it could get more charm.

As the Inputs (Multipliers) for the matrix could be:
- LFO ( Sinus from -0.5 to 0.5, add 1 to get in to the range of 0 to 1)
- MOD-Wheel 
- Velocity

Output?
- Velocity
- Pitch
- ADSR-Attack ?
- Filter-Attack ?
- Filter-Freq
- Filter-Q

**Comment by marcel-licence** [comment]
I used the modulation matrix in the Roland SonicCell a lot. It is really a nice feature and allows to be much more creative. I won't miss it.

I am not sure how much processing will be done finally. Commercial products usually use DSP's beeing much more faster and I guess they also using very high optimized code. 

There would be also a good interface necessary to avoid getting lost in the matrix xD

**Comment by ErichHeinemann** [comment]
Indeed, Marcel, You are right. The Interface would be a problem.

Then shift this to a later version.

Best Regards
Erich

**Comment by GillesLACAUD** [comment]
Do not forget the AFTERTOUCH as a mod source, the Pitch bend can also be route as a mod source.
For the output you can also add (if you have two LFOs) Lfo amount and lfo speed

**Comment by GillesLACAUD** [comment]
For the interface and without add any hardware i think you can maybe use the serial debug com as a VT100 console not only for output but also for input. Another option is to use the Bluetooth or Wifi part...but i think we will lost a lot of performance for the audio part. As you know my option is to use an Nextion touch screen control by an uart. You have differents size and options for these screens

**Comment by GillesLACAUD** [comment]
For the matrix it is not so complicated if you just assign the source to one destination and with one amount. It is more complicate if you choose multiple destinations with different amounts.

**Comment by GillesLACAUD** [comment]
Add of course add the PWM to Matrix destination...
